### PR TITLE
Teach _search to read search time mappings

### PR DESCRIPTION
**Pull request by nik9000**
This adds support for search time mappings to the _search API like so:
```
curl -XPOST -HContent-Type:application/json -uelastic:password 'localhost:9200/test/_search?pretty&error_trace' -d'{
  "runtime_mappings": {
    "cat": {
      "type": "script",
      "runtime_type": "keyword",
      "script": "value(\"cat\")"
    }
  },
  "docvalue_fields": ["cat"]
}'
```

It just lifts the `mappings` field from the create index and mappings
update requests and drops it into the `_search` request. It then gives
each mapped field a chance to convert itself into a runtime field. Most
mapping can not and will throw an exception that is translated into an
HTTP 400 response. The `script` converts itself properly and is
available at runtime.

Relates to #59332

### CODE REVIEW

**Comment by nik9000** [comment]
>   "mappings": {
>     "properties": {

@javanna and I talked about changing it to just `runtime_mappings`. I used this syntax here because it is convenient, but it is, like the TODO in the code says, too big.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 1]
```diff
@@ -618,6 +618,38 @@ public Analyzer searchQuoteAnalyzer() {
...
+     * any {@code extraMapping} provided.
+     * @param extraMapping extra mappings parse and to add to the request
+     *        lookup or {@code null} if there aren't any extra mappings  
+     */
+    public Function<String, MappedFieldType> newFieldTypeLookup(Map<String, Object> extraMapping) {
+        if (extraMapping == null || extraMapping.size() == 0) {
+            return this::fieldType;
+        }
+        // TODO parsing an entire mapping is almost certainly "too big"
+        Mapping mapping = documentMapperParser().parseMapping("extra", extraMapping);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:632]: I am not sure about parsing the whole thing as mappings. I would not require `properties` to start with and I think I would parse each type manually, e.g. 

```
Mapper.TypeParser.ParserContext parserContext = mapperService.documentMapperParser().parserContext();
Mapper.TypeParser typeParser = parserContext.typeParser(type);
```

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 2]
```diff
@@ -618,6 +618,38 @@ public Analyzer searchQuoteAnalyzer() {
...
+            for (MappedFieldType ft: m.convertToSearchTimeMappings()) {
+                MappedFieldType fromIndexMapping = fieldType(ft.name());
+                if (fromIndexMapping != null) {
+                    throw new IllegalArgumentException("No shadowing?! Was [" + fromIndexMapping + "] attempted to add [" + mapping + "]");
+                }
+                extra.put(ft.name(), ft);
+            }
+        }
+        return fullName -> {
+            MappedFieldType searchTime = extra.get(fullName);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:645]: do we want to give the precedence to ephemeral runtime fields if they are defined with the same name as a mapped field? Or do we want to reject the search?

File: server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 3]
```diff
@@ -241,7 +265,7 @@ public MappedFieldType fieldMapper(String name) {
         if (name.equals(TypeFieldMapper.NAME)) {
             deprecationLogger.deprecate("query_with_types", TYPES_DEPRECATION_MESSAGE);
         }
-        return failIfFieldMappingNotFound(name, mapperService.fieldType(name));
+        return failIfFieldMappingNotFound(name, lookupFieldType.apply(name));
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:268]: I am not sure I see the value in using a different function from mapper service. If the function is only used here, can't we have the logic here without a function and call the usual function for the ordinary mapped fields?

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java [hunk 4]
```diff
@@ -775,7 +776,7 @@ public FetchSearchResult fetchResult() {
 
     @Override
     public MappedFieldType fieldType(String name) {
-        return mapperService().fieldType(name);
+        return queryShardContext.fieldMapper(name);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:779]: I had missed this in my POC, which I guess made the fetch phase not work but I never tested that.

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 5]
```diff
@@ -895,6 +910,21 @@ public SearchSourceBuilder stats(List<String> statsGroups) {
         return stats;
     }
 
+    /**
+     * Add a search time mapping.
+     */
+    public SearchSourceBuilder extraMapping(Map<String, Object> extraMapping) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:916]: runtimeMappings?

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 6]
```diff
@@ -1104,6 +1135,8 @@ public void parseXContent(XContentParser parser, boolean checkTrailingTokens) th
                     sliceBuilder = SliceBuilder.fromXContent(parser);
                 } else if (COLLAPSE.match(currentFieldName, parser.getDeprecationHandler())) {
                     collapse = CollapseBuilder.fromXContent(parser);
+                } else if (CreateIndexRequest.MAPPINGS.match(currentFieldName, parser.getDeprecationHandler())) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1138]: I would rename the section to runtime_mappings, at least for now. I think it better defines that it only holds mappings for runtime fields

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java [hunk 7]
```diff
@@ -71,7 +71,7 @@ public void hitsExecute(SearchContext context, SearchHit[] hits) throws IOExcept
 
         for (FieldAndFormat fieldAndFormat : context.docValuesContext().fields()) {
             String field = fieldAndFormat.field;
-            MappedFieldType fieldType = context.mapperService().fieldType(field);
+            MappedFieldType fieldType = context.fieldType(field);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchDocValuesPhase.java:74]: ++

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 8]
```diff
@@ -172,4 +173,11 @@ public final String simpleName() {
      *  Both {@code this} and {@code mergeWith} will be left unmodified. */
     public abstract Mapper merge(Mapper mergeWith);
 
+    /**
+     * Called when this {@linkplain Mapper} is parsed on the {@code _search}
+     * request to make fields available just for this search. 
+     */
+    public List<MappedFieldType> convertToSearchTimeMappings() {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:180]: could this rather be a boolean isRuntimeField ? I see though that you are also resolving a Mapper into a list of MappedFieldTypes, but why a list? are you thinking of grok that will expand to multiple fields?

File: server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 9]
```diff
@@ -119,18 +121,38 @@ public QueryShardContext(int shardId,
...
+            nowInMillis,
+            indexNameMatcher,
+            new Index(
+                RemoteClusterAware.buildRemoteIndexName(clusterAlias, indexSettings.getIndex().getName()),
+                indexSettings.getIndex().getUUID()
+            ),
+            allowExpensiveQueries,
+            valuesSourceRegistry,
+            mapperService.newFieldTypeLookup(extraMappings)
+        );
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:148]: although I see the value in reformatting like this as we go, I would try not to do it in a feature branch, because it will cause merge conflicts. Though I see how in this particular case, changing arguments here will anyway cause merge conflicts with master because things are not formatted nicely like you did.

**Review by javanna — COMMENTED** [review]
I left some comments, I guess one thing to look at is whether we could parse/validate the new section in the coordinating node so that we could fail early and once if something is wrong or if a non-runtime field is defined as part of the search request. Could we also validate the script once? I'm afraid it will be tough because MapperService is an index component, but there may be other ways to only validate?

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 10]
```diff
@@ -239,6 +244,9 @@ public SearchSourceBuilder(StreamInput in) throws IOException {
         sliceBuilder = in.readOptionalWriteable(SliceBuilder::new);
         collapse = in.readOptionalWriteable(CollapseBuilder::new);
         trackTotalHitsUpTo = in.readOptionalInt();
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:247]: :+1:

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java [hunk 11]
```diff
@@ -293,6 +301,13 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeOptionalWriteable(sliceBuilder);
         out.writeOptionalWriteable(collapse);
         out.writeOptionalInt(trackTotalHitsUpTo);
+        if (out.getVersion().onOrAfter(Version.V_8_0_0)) {
+            out.writeMap(extraMapping);
+        } else {
+            throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:307]: If we don't write the map then nodes that can't understand will silently ignore the fields which feels bad!

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:632]: Me neither! I'll switch it to what you say. I thought what I had was a good start for us to talk about, but Mapping is so big!

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 2 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:645]: I've been rejecting them at parse time. But I could see the argument to take preference.

File: server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 3 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:268]: I like the Function because it forces the parsing and lookup code close together and makes testing pretty easy. I'm certainly open to other ideas but I had the parsing here at first and didn't like it. I'd be happy to move the function's creation to a static method if you'd prefer that. That way I can still test it without building all the stuff to make a whole QueryShardContext.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java (hunk 4 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:779]: Yeah! There are a bunch of places that use the MapperService on QueryShardContext that aren't going to work well. I haven't thought through a good way to deal with that.

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 5 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:916]: Sure. I think I prefer "searchTime" but we've been calling them runtime for a while now.

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 6 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:1138]: I guess. It is in the search request. But if we're already going to drop the need for `properties` then trying to line it up with the index creation api isn't really right anyway.

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java (hunk 8 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:180]: I had it this way for grok and for the other things I was experimenting with. And because `Mapper` doesn't have a MappedFieldType on it, only FieldMapper. But I think changing how we parse will make this less of a thing. If you'd prefer to start with a boolean I'm happy to do that and change it for grok if we need to.

File: server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 9 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:148]: That was just a habit from the work I and been doing. I hadn't been thinking about the branch. I'll have another look on Monday.

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:307]: oh true, so only some nodes may have the runtime fields which messes up the results? This is pretty important indeed. I wonder what effect does throwing here makes from a user's point of view.

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 2 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:645]: Rejecting is good, I did not realize that that is what you did.

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java (hunk 8 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:180]: yes please I would prefer not to worry about grok in this PR. the cast Mapper to FieldMapper is odd indeed, maybe we can address that upstream.

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:307]: also is not completely the truth. it depends on the nodes that get hit?

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 12]
```diff
@@ -618,6 +618,38 @@ public Analyzer searchQuoteAnalyzer() {
...
+        }
+        // TODO parsing an entire mapping is almost certainly "too big"
+        Mapping mapping = documentMapperParser().parseMapping("extra", extraMapping);
+        Map<String, MappedFieldType> extra = new HashMap<>();
+        for (Mapper m : mapping.root()) {
+            // TODO this should dig the sub fields out of object mappers. Probably.
+            for (MappedFieldType ft: m.convertToSearchTimeMappings()) {
+                MappedFieldType fromIndexMapping = fieldType(ft.name());
+                if (fromIndexMapping != null) {
+                    throw new IllegalArgumentException("No shadowing?! Was [" + fromIndexMapping + "] attempted to add [" + mapping + "]");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:639]: maybe reword the error to make it more user-friendly?

File: server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:307]: The last I checked it'll cause the request to `400` with the message. It does depend on which node gets hit, yeah.

I think it'll also come up in CCS where I expect the remote cluster could be at an older version for longer. I'll update the error message a little.

**Comment by nik9000** [comment]
Just added an integration test! Yay!

File: server/src/main/java/org/elasticsearch/index/IndexService.java [hunk 13]
```diff
@@ -584,14 +584,25 @@ public IndexSettings getIndexSettings() {
...
         final SearchIndexNameMatcher indexNameMatcher =
             new SearchIndexNameMatcher(index().getName(), clusterAlias, clusterService, expressionResolver);
         return new QueryShardContext(
             shardId, indexSettings, bigArrays, indexCache.bitsetFilterCache(), indexFieldData::getForField, mapperService(),
             similarityService(), scriptService, xContentRegistry, namedWriteableRegistry, client, searcher, nowInMillis, clusterAlias,
-            indexNameMatcher, allowExpensiveQueries, valuesSourceRegistry);
+            indexNameMatcher, allowExpensiveQueries, valuesSourceRegistry, runtimeMappings);
     }
 
+
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/IndexService.java:605]: remove empty line? :)

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java [hunk 14]
```diff
@@ -171,5 +171,4 @@ public final String simpleName() {
     /** Return the merge of {@code mergeWith} into this.
      *  Both {@code this} and {@code mergeWith} will be left unmodified. */
     public abstract Mapper merge(Mapper mergeWith);
-
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:174]: revert?

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 15]
```diff
@@ -618,6 +618,65 @@ public Analyzer searchQuoteAnalyzer() {
         return this.searchQuoteAnalyzer;
     }
 
+    /**
+     * Builds a {@linkplain Function} to lookup mappers for a request, adding
+     * any {@code extraMapping} provided.
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:623]: runtime mappings?

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java [hunk 16]
```diff
@@ -618,6 +618,65 @@ public Analyzer searchQuoteAnalyzer() {
...
+                throw new IllegalArgumentException("[type] is required for runtime mapping [" + runtimeEntry.getKey() + "]");
+            }
+            Mapper.TypeParser parser = documentMapperParser().parserContext().typeParser(type);
+            if (parser == null) {
+                throw new IllegalArgumentException("[" + type + "] is unknown type for runtime mapping [" + runtimeEntry.getKey() + "]");
+            }
+            Mapper.Builder<?> builder = parser.parse(runtimeEntry.getKey(), definition, documentMapperParser().parserContext());
+            Mapper mapper = builder.build(builderContext);
+            
+            MapperUtils.collect(mapper, objectMappers, fieldMappers, fieldAliasMappers);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:649]: Can you expand on why we need this? I was expecting that alias or objects would be caught anyways by your checks below, that only runtime fields can be defined.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java (hunk 4 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:779]: do you think that we should dig on those usages before merging this PR? I am still wondering if exposing runtime mappings this way is enough, I hope it would be but then why does query shard context expose the mapper service directly? Maybe we also have a chance to clean this up upstream, not sure.

**Review by javanna — COMMENTED** [review]
I left some comments/questions

File: server/src/main/java/org/elasticsearch/index/IndexService.java (hunk 13 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/IndexService.java:605]: :heart:

File: server/src/main/java/org/elasticsearch/index/mapper/Mapper.java (hunk 14 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/Mapper.java:174]: :+1:

File: server/src/main/java/org/elasticsearch/index/mapper/MapperService.java (hunk 15 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:623]: Damn, I missed it.

File: server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java (hunk 4 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:779]: I looked into this some more and I think I really have to move this change into `MapperService` and build a new mapper service that is aware of the new fields. Even then weird things are likely to slip through, but fewer of them. We just rely on stuff like `getMapperService().searchAnalyzer()` and stuff like that which is all really tightly. I'll dig further.

**Comment by javanna** [comment]
I marked this stalled as we decided to postpone allowing to define runtime fields as part of the search request. We will get back to this later.
